exporters/prometheus/src/test/java/io/opentelemetry/exporter/prometheus/PrometheusUnitsHelperTest.java

```diff
@@ -1,108 +0,0 @@
-        Arguments.of("d", "days"),
-        // Simple expansion - Time unit h
-        Arguments.of("h", "hours"),
-        // Simple expansion - Time unit s
-        Arguments.of("s", "seconds"),
-        // Simple expansion - Time unit ms
-        Arguments.of("ms", "milliseconds"),
-        // Simple expansion - Time unit us
-        Arguments.of("us", "microseconds"),
-        // Simple expansion - Time unit ns
-        Arguments.of("ns", "nanoseconds"),
-        // Simple expansion - Time unit min
-        Arguments.of("min", "minutes"),
-        // Simple expansion - special symbol - %
-        Arguments.of("%", "percent"),
-        // Simple expansion - frequency
-        Arguments.of("Hz", "hertz"),
-        // Simple expansion - temperature
-        Arguments.of("Cel", "celsius"),
-        // Unit not found - Case sensitive
-        Arguments.of("S", "S"),
-        // Special case - 1
-        Arguments.of("1", ""),
-        // Special Case - Drop metric units in {}
-        Arguments.of("{packets}", ""),
-        // Special Case - Dropped metric units only in {}
-        Arguments.of("{packets}V", "volts"),
-        // Special Case - Dropped metric units with 'per' unit handling applicable
-        Arguments.of("{scanned}/{returned}", ""),
-        // Special Case - Dropped metric units with 'per' unit handling applicable
-        Arguments.of("{objects}/s", "per_second"),
-        // Units expressing rate - 'per' units, both units expanded
-        Arguments.of("m/s", "meters_per_second"),
-        // Units expressing rate - per minute
-        Arguments.of("m/m", "meters_per_minute"),
-        // Units expressing rate - per day
-        Arguments.of("A/d", "amperes_per_day"),
-        // Units expressing rate - per week
-        Arguments.of("W/w", "watts_per_week"),
-        // Units expressing rate - per month
-        Arguments.of("J/mo", "joules_per_month"),
-        // Units expressing rate - per year
-        Arguments.of("TBy/y", "terabytes_per_year"),
-        // Units expressing rate - 'per' units, both units unknown
-        Arguments.of("v/v", "v_per_v"),
-        // Units expressing rate - 'per' units, first unit unknown
-        Arguments.of("km/h", "km_per_hour"),
-        // Units expressing rate - 'per' units, 'per' unit unknown
-        Arguments.of("g/g", "grams_per_g"),
-        // Misc - unit containing known abbreviations improperly formatted
-        Arguments.of("watts_W", "watts_W"),
-        // Unsupported symbols
-        Arguments.of("°F", "F"),
-        // Unsupported symbols - multiple
-        Arguments.of("unit+=.:,!* & #unused", "unit_unused"),
-        // Unsupported symbols - 'per' units
-        Arguments.of("__test $/°C", "test_per_C"),
-        // Unsupported symbols - whitespace
-        Arguments.of("\t", ""),
-        // Null unit
-        Arguments.of(null, null),
-        // Misc - unit cleanup - no case match special char
-        Arguments.of("$1000", "1000"));
```

Comment:
Can we adapt / retain this test to confirm the new behavior of `PrometheusUnitHelper`?

Reply:
Good point. I restored the test, but I also changed some methods where I think the test is wrong:

* `Arguments.of("m/m", "meters_per_minute")` -> `Arguments.of("m/min", "meters_per_minute")`
* `Arguments.of("W/w", "watts_per_week"),` -> `Arguments.of("W/wk", "watts_per_week")`
* `Arguments.of("TBy/y", "terabytes_per_year")` -> `Arguments.of("TBy/a", "terabytes_per_year")`
* `Arguments.of("g/g", "grams_per_g")` -> `Arguments.of("g/x", "grams_per_x")` (this is supposed to test an unknown "per" unit, but `g` is not unknown
* Removed all tests for removing special characters like `Arguments.of("__test $/°C", "test_per_C")` because I could not find any reference that these characters should be removed.

Please let me know if I got it wrong.

Reply:
This seems reasonable to me. @psx95 please take a look if you can.

Reply:
I just wanted to confirm the following - 
 - In the [collector-contrib](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/aaaaa0c5555cab8388ab523e5ae107cc8f0d235a/pkg/translator/prometheus/normalize_name.go#L58C5-L58C15), there is a separate list of 'per' unit(s). Which is why the "per" test cases were using those units specifically.
  - I do not see "wk" as a specified unit in the collector. Are we okay with the discrepancy between collector and SDK ? 
  - In the collector, certain units - for instance "mo", "y" are only used for "per" unit cases i.e. only in the second part of the metric unit when split using "/".  IIUC, the current implementation might use the mappings for any part of the unit. Is this intentional ?
   - Regarding the removing unsupported symbols - it was from the [OTel Spec Doc, OTLP -> Prometheus](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/compatibility/prometheus_and_openmetrics.md#metric-metadata-1). It mentions that metric name is supposed to match the regex `[a-zA-Z_:]([a-zA-Z0-9_:])*` which makes symbols like `$` invalid and replaced with `_` instead.

Reply:
> I do not see "wk" as a specified unit in the collector. Are we okay with the discrepancy between collector and SDK ?

While I like the idea of symmetry with the collector, the `wk` unit is defined in [ucum](https://unitsofmeasure.org/ucum), so I'm ok including it. UCUM is defined in the [metric semantic conventions](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/metrics.md#instrument-units) for defining unit.

> In the [collector-contrib](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/aaaaa0c5555cab8388ab523e5ae107cc8f0d235a/pkg/translator/prometheus/normalize_name.go#L58C5-L58C15), there is a separate list of 'per' unit(s). Which is why the "per" test cases were using those units specifically.
> In the collector, certain units - for instance "mo", "y" are only used for "per" unit cases i.e. only in the second part of the metric unit when split using "/". IIUC, the current implementation might use the mappings for any part of the unit. Is this intentional ?

The specification doesn't include any language that restricts the conversion to the `*_per_*` version to a subset of units, so I think its fine. Makes it simpler to understand.

> Regarding the removing unsupported symbols - it was from the [OTel Spec Doc, OTLP -> Prometheus](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/compatibility/prometheus_and_openmetrics.md#metric-metadata-1). It mentions that metric name is supposed to match the regex [a-zA-Z_:]([a-zA-Z0-9_:])* which makes symbols like $ invalid and replaced with _ instead.

Metric names are still sanitized according to those rules. However, we now delegate to the prometheus client library: https://github.com/open-telemetry/opentelemetry-java/blob/main/exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/Otel2PrometheusConverter.java#L448

> In places where the spec was ambiguous, the unit conversion implementation in https://github.com/open-telemetry/opentelemetry-java/pull/5400 was done to be consistent with the collector implementation.

If we've diverged in any ways that aren't discussed here, I'd love to resolve those inconsistencies.

Reply:
I see, thanks for clearing this up. 

One thing I did notice was that the definition of illegal characters is different in [Prometheus Client Library](https://github.com/prometheus/client_java/blob/60f72b6d27d31a4a566670cebdfb2699073c705a/prometheus-metrics-model/src/main/java/io/prometheus/metrics/model/snapshots/PrometheusNaming.java#L16) and the regex defined in [OTel Spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/compatibility/prometheus_and_openmetrics.md#otlp-metric-points-to-prometheus).  (*OTel spec does not allow `.`*)
 - Regex allowed by OTel Spec - `[a-zA-Z_:]([a-zA-Z0-9_:])*`
 - Regex allowed by Prometheus Client Library - `^[a-zA-Z_.:][a-zA-Z0-9_.:]+$`

Reply:
> If we've diverged in any ways that aren't discussed here, I'd love to resolve those inconsistencies.

I think we should add back the tests that were testing for removal of special characters and leading/trailing `_` in the unit. 

Regarding special characters: While it is not explicitly specified in spec, it is implicitly indicated by the regex used in the prometheus client library.

Regarding leading/trailing `_`: Removing these are not mandated by the spec, but the implementation of unit conversion in the collector does not allow for any leading or trailing `_`.

Reply:
> illegal characters

Here's the context: The Prometheus community decided (see [dev summit 2023-09](https://docs.google.com/document/d/11LC3wJcVk00l8w5P3oLQ-m3Y37iom6INAMEu2ZAGIIE/edit#heading=h.np6zum9h19xx)) to allow UTF-8 characters in metric and attribute names for better OTel compatibility.

This is not implemented in the Prometheus server yet, so today the Prometheus exposition format will only create metric names and attribute names matching `[a-zA-Z_:]([a-zA-Z0-9_:])*` (without the dot).

However, we extended the Prometheus client library to allow dots in metric names, hence the new regexp that allows dots.

The idea is: Today the dots will never be exposed, they are implicitly converted to underscores in the exposition format. However, users can use dots when creating metrics, so they can use metric names compliant with OpenTelemetry's semantic conventions today. As soon as the Prometheus server and exposition format supports dots in metric names they can enable dots without having to re-instrument their applications.

TL;DR: While dots are allowed when creating metrics, they are implicitly replaced with underscores when exposing metrics, at least until dots become officially supported by the Prometheus server and exposition format.

Reply:
Thanks for the context @fstab.